### PR TITLE
[FLINK-28664][python][format] FileSink supports AvroWriters

### PR DESCRIPTION
**Pull request by Vancior**
## What is the purpose of the change

This PR supports using `AvroWriters.for_generic_record` to write vanilla Python objects to files via FileSink.

## Brief change log

- add Avro encoder on Python side
- port Java AvroWriters.forGenericRecord to Python

## Verifying this change

This change added tests and can be verified as follows:

- `FileSinkAvroWritersTests` in test_file_system.py

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs & Python Sphinx doc)

### CODE REVIEW

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/avro/AvroParquetReaders.java [hunk 1]
```diff
@@ -51,15 +51,7 @@ public class AvroParquetReaders {
     public static <T extends SpecificRecordBase> StreamFormat<T> forSpecificRecord(
             final Class<T> typeClass) {
         return new AvroParquetRecordFormat<>(
-                new AvroTypeInfo<>(typeClass),
```
Comment (dianfu) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/avro/AvroParquetReaders.java:54]: It seems that you are reverting the changes made in #19956. Are there any specify reasons?

File: flink-python/pyflink/datastream/formats/csv.py [hunk 2]
```diff
@@ -131,7 +131,7 @@ def add_columns_from(self, schema: 'CsvSchema') -> 'CsvSchemaBuilder':
         :param schema: Another :class:`CsvSchema`.
         """
         self._j_schema_builder.addColumnsFrom(schema._j_schema)
-        for field in cast(schema._data_type, RowType):
```
Comment (dianfu) [inline flink-python/pyflink/datastream/formats/csv.py:134]: Could we update the test cases to cover this method?

File: flink-python/pyflink/datastream/formats/avro.py [hunk 3]
```diff
@@ -63,6 +70,9 @@ class GenericRecordAvroTypeInfo(TypeInformation):
     A :class:`TypeInformation` of Avro's GenericRecord, including the schema. This is a wrapper of
     Java org.apache.flink.formats.avro.typeutils.GenericRecordAvroTypeInfo.
```
Comment (dianfu) [inline flink-python/pyflink/datastream/formats/avro.py:71]: This class is also a public class which will be used by users and so add it to the __init__.py to expose it?

**Review by dianfu — COMMENTED** [review]
@Vancior Thanks a lot for this PR. Good work! Have left a few minor comments. Looking forward to the update!

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/avro/AvroParquetReaders.java (hunk 1 above)
Comment (Vancior) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/avro/AvroParquetReaders.java:54]: Yeah I find that the change is only needed for the GenericRecord we provided in Python.

File: flink-python/pyflink/datastream/formats/csv.py (hunk 2 above)
Comment (Vancior) [reply flink-python/pyflink/datastream/formats/csv.py:134]: I could add this in the CSV PR.

File: flink-python/pyflink/datastream/formats/avro.py (hunk 3 above)
Comment (Vancior) [reply flink-python/pyflink/datastream/formats/avro.py:71]: Quite right, I'll update it.

File: flink-python/pyflink/datastream/connectors/file_system.py [hunk 4]
```diff
@@ -535,6 +568,51 @@ def for_row_format(base_path: str, encoder: Encoder) -> 'FileSink.RowFormatBuild
         return FileSink.RowFormatBuilder(
             JFileSink.forRowFormat(JPath(base_path), encoder._j_encoder))
 
+    class BulkFormatBuilder(object):
+        """
+        Builder for the vanilla FileSink using a bulk format.
+        """
+
+        def __init__(self, j_bulk_format_builder):
```
[the remark this answers is not in the document]
Comment (Vancior) [reply flink-python/pyflink/datastream/connectors/file_system.py:576]: Maybe we can separate it to another PR? I'm just aligning to the builder in StreamingFileSink here.

[the remark this answers is not in the document]
Comment (dianfu) [reply flink-python/pyflink/datastream/connectors/file_system.py:576]: StreamingFileSink is to be removed, see https://issues.apache.org/jira/browse/FLINK-28641 for more details. Will we align these methods in 1.16?

**Pull request closed without merging**